File: java/src/main/java/lab/gabon/service/TokenStore.java
package lab.gabon.service;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.TimeUnit;
import lab.gabon.config.AppConfig;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.data.redis.core.script.RedisScript;
import org.springframework.stereotype.Service;

/**
 * Redis-backed token store for JWT blacklisting and refresh-token family tracking.
 *
 * <p>Key layout:
 *
 * <ul>
 *   <li>{@code token:blacklist:{jti}} -- value "1", TTL = access token remaining lifetime
 *   <li>{@code token:family:{familyId}} -- value "{userId}:{currentJti}", TTL = refresh token
 *       lifetime
 * </ul>
 */
@Service
public class TokenStore {

  /** Result of a compare-and-swap on a refresh token family. */
  public sealed interface CasResult {
    /** CAS succeeded; the family now points to the new JTI. */
    record Success(long userId) implements CasResult {}

    /** The family key does not exist (expired or deleted). */
    record Missing() implements CasResult {}

    /** The current JTI did not match (replay detected). Family was deleted. */
    record Conflict() implements CasResult {}
  }

  private static final long CAS_MISSING = -1L;
  private static final long CAS_CONFLICT = -2L;

  private final StringRedisTemplate redis;
  private final RedisScript<Long> casScript;
  private final AppConfig.JwtConfig jwtConfig;

  public TokenStore(
      StringRedisTemplate redis,
      @Qualifier("refreshCasScript") RedisScript<Long> casScript,
      AppConfig appConfig) {
    this.redis = redis;
    this.casScript = casScript;
    this.jwtConfig = appConfig.jwt();
  }

  // -- Refresh Token Family ---------------------------------------------------

  /** Create or overwrite a token family entry with the appropriate refresh TTL. */
  public void storeRefreshFamily(String familyId, long userId, String refreshJti, boolean admin) {
    var ttl = admin ? jwtConfig.adminRefreshTtl() : jwtConfig.customerRefreshTtl();
    var key = "token:family:" + familyId;
    var value = userId + ":" + refreshJti;
    redis.opsForValue().set(key, value, ttl.toSeconds(), TimeUnit.SECONDS);
  }

  /**
   * Atomically compare-and-swap the current JTI in a token family.
   *
   * @return userId on success, or Missing/Conflict result
   */
  public CasResult rotateRefreshToken(String familyId, String oldJti, String newJti) {
    var key = "token:family:" + familyId;
    var result = redis.execute(casScript, List.of(key), oldJti, newJti);
    if (result == null || result == CAS_MISSING) {
      return new CasResult.Missing();
    }
    if (result == CAS_CONFLICT) {
      return new CasResult.Conflict();
    }
    return new CasResult.Success(result);
  }

  /** Delete an entire token family (e.g. on logout). */
  public void revokeFamily(String familyId) {
    redis.delete("token:family:" + familyId);
  }

  // -- Access Token Blacklist -------------------------------------------------

  /** Add a JTI to the blacklist with the appropriate access TTL. */
  public void blacklistAccessToken(String jti, boolean admin) {
    Duration ttl = admin ? jwtConfig.adminAccessTtl() : jwtConfig.customerAccessTtl();
    var key = "token:blacklist:" + jti;
    redis.opsForValue().set(key, "1", ttl.toSeconds(), TimeUnit.SECONDS);
  }

  /** Check whether a JTI has been blacklisted. */
  public boolean isBlacklisted(String jti) {
    return Boolean.TRUE.equals(redis.hasKey("token:blacklist:" + jti));
  }
}

File: kotlin/src/main/kotlin/lab/gabon/service/TokenStore.kt
package lab.gabon.service

import io.lettuce.core.ExperimentalLettuceCoroutinesApi
import io.lettuce.core.ScriptOutputType
import io.lettuce.core.api.coroutines.RedisCoroutinesCommands

/** Result of a compare-and-swap on a refresh token family. */
sealed interface CasResult {
    /** CAS succeeded; the family now points to the new JTI. */
    data class Success(
        val userId: Long,
    ) : CasResult

    /** The family key does not exist (expired or deleted). */
    data object Missing : CasResult

    /** The current JTI did not match the expected value (replay detected). Family was deleted. */
    data object Conflict : CasResult
}

/**
 * Redis-backed token store for JWT blacklisting and refresh-token family tracking.
 *
 * Key layout:
 * - `token:blacklist:{jti}` — value `"1"`, TTL = access token remaining lifetime
 * - `token:family:{familyId}` — value `"{userId}:{currentJti}"`, TTL = refresh token lifetime
 */
@OptIn(ExperimentalLettuceCoroutinesApi::class)
class RedisTokenStore(
    private val redis: RedisCoroutinesCommands<String, String>,
) {
    // ── Blacklist ──────────────────────────────────────────────

    /** Add [jti] to the blacklist with the given TTL in seconds. */
    suspend fun setBlacklist(
        jti: String,
        ttlSeconds: Long,
    ) {
        redis.setex("token:blacklist:$jti", ttlSeconds, "1")
    }

    /** Check whether [jti] has been blacklisted. */
    suspend fun isBlacklisted(jti: String): Boolean = (redis.exists("token:blacklist:$jti") ?: 0L) > 0L

    // ── Refresh Token Family ───────────────────────────────────

    /** Create or overwrite a token family entry. */
    suspend fun setFamily(
        familyId: String,
        userId: Long,
        currentJti: String,
        ttlSeconds: Long,
    ) {
        redis.setex("token:family:$familyId", ttlSeconds, "$userId:$currentJti")
    }

    /**
     * Atomically compare-and-swap the current JTI in a token family.
     *
     * Lua script behaviour:
     * - Key missing → return `-1` → [CasResult.Missing]
     * - Current JTI doesn't match [expectedJti] → delete key, return `-2` → [CasResult.Conflict]
     * - Match → update value with [newJti] (KEEPTTL), return `userId` → [CasResult.Success]
     */
    suspend fun casFamily(
        familyId: String,
        expectedJti: String,
        newJti: String,
    ): CasResult {
        val result =
            redis.eval<Long>(
                CAS_SCRIPT,
                ScriptOutputType.INTEGER,
                arrayOf("token:family:$familyId"),
                expectedJti,
                newJti,
            )
        return when {
            result == null || result == CAS_MISSING -> CasResult.Missing
            result == CAS_CONFLICT -> CasResult.Conflict
            else -> CasResult.Success(userId = result)
        }
    }

    /** Delete an entire token family (e.g. on logout). */
    suspend fun deleteFamily(familyId: String) {
        redis.del("token:family:$familyId")
    }

    private companion object {
        const val CAS_MISSING = -1L
        const val CAS_CONFLICT = -2L

        /**
         * Lua CAS script for refresh token rotation.
         *
         * KEYS[1] = token:family:{familyId}
         * ARGV[1] = expectedJti
         * ARGV[2] = newJti
         *
         * Returns:
         *   userId (positive Long) on success
         *   -1 if key is missing
         *   -2 if JTI mismatch (replay attack — key is deleted)
         */
        val CAS_SCRIPT =
            """
            local v = redis.call('GET', KEYS[1])
            if v == false then return -1 end
            local sep = string.find(v, ':', 1, true)
            local uid = string.sub(v, 1, sep - 1)
            local jti = string.sub(v, sep + 1)
            if jti ~= ARGV[1] then
              redis.call('DEL', KEYS[1])
              return -2
            end
            redis.call('SET', KEYS[1], uid .. ':' .. ARGV[2], 'KEEPTTL')
            return tonumber(uid)
            """.trimIndent()
    }
}
